File: spriteforest-html5/base/SFNode.js
;
/*!
 *
 *  SpriteForest JavaScript Library v1.0.0
 *  http://github.com/moky/SpriteForest/js/
 *
 *  Copyright (c) 2014 moKy at slanissue.com
 *  Released under the MIT license
 *  http://en.wikipedia.org/wiki/MIT_License
 *
 */
//
//  SFNode.js
//  SpriteForest-HTML5
//
//  Created by moKy on Feb. 26, 2014.
//

!function(sf) {
	'use strict';
	var cc = sf.cc;
	var cn = sf.cn;
	
	//
	//  construtor
	//
	var Node = function() {
		cn.Node.prototype.ctor.call(this);
	};
	
	var nodeWithString = function(string) {
		var Class = cn.classFromString(string);
		if (typeof Class !== 'function') {
			cn.error(string);
			return null;
		}
		
		var node = new Class();
		if (typeof node.init === 'function') {
			node.init();
		}
		return node;
	};
	var nodeWithDictionary = function(dict) {
		var Class = cn.classFromString(dict['Class']);
		if (typeof Class !== 'function') {
			cn.error(dict);
			return null;
		}
		
		var node = new Class();
		if (typeof node.initWithDictionary === 'function') {
			node.initWithDictionary(dict);
		} else if (typeof node.init === 'function') {
			node.init();
		}
		return node;
	};
	
	//
	//  static functions
	//
	var create = function() {
		if (arguments.length === 0 || !arguments[0]) {
			var node = new sf.Node();
			node.init();
			return node;
		} else if (typeof arguments[0] === 'string') {
			return nodeWithString(arguments[0]);
		} else if (typeof arguments[0] === 'object') {
			return nodeWithDictionary(arguments[0]);
		} else {
			cn.error(arguments);
		}
		return null;
	};
	
	//
	//  instance functions
	//
	
	var _standard_string = function(string) {
		return string ? string.replace(/\s/g, '').toLowerCase() : '';
	};
	
	var processPositionAdjustment = function(child, dict, parent) {
		if (!child || child.getPositionX() == 0 || child.getPositionY() == 0) {
			return;
		}
		
		// standard string
		var position = _standard_string(dict['Position']);
		if (!position || position.length === 0) {
			return;
		} else if (position === 'topleft') {
			position = '{left,top}';
		} else if (position === 'topright') {
			position = '{right,top}';
		} else if (position === 'center') {
			position = '{center,middle}';
		} else if (position === 'bottomleft') {
			position = '{left,bottom}';
		} else if (position === 'bottomright') {
			position = '{right,bottom}';
		}
		
		// replace values
		if (position.length >= 7) { // "{0,top}"
			var parentSize = parent.getContentSize();
			
			var top = 0,
			left = 0,
			bottom = parentSize.height,
			right = parentSize.width,
			middle = bottom * 0.5,
			center = right * 0.5;
			
			var pos;
			// top
			pos = position.indexOf('top');
			if (pos >= 0)  {
				position = position.substr(0, pos) + top + position.substr(pos + 3);
			}
			// left
			pos = position.indexOf('left');
			if (pos >= 0)  {
				position = position.substr(0, pos) + left + position.substr(pos + 4);
			}
			// bottom
			pos = position.indexOf('bottom');
			if (pos >= 0)  {
				position = position.substr(0, pos) + bottom + position.substr(pos + 6);
			}
			// right
			pos = position.indexOf('right');
			if (pos >= 0)  {
				position = position.substr(0, pos) + right + position.substr(pos + 5);
			}
			// middle
			pos = position.indexOf('middle');
			if (pos >= 0)  {
				position = position.substr(0, pos) + middle + position.substr(pos + 6);
			}
			// center
			pos = position.indexOf('center');
			if (pos >= 0)  {
				position = position.substr(0, pos) + center + position.substr(pos + 6);
			}
		}
		
		// set it
		if (position.length >= 5) { // "{0,0}"
			child.setPosition(cc.PointFromString(position));
		}
	};
	
	var processScaleMode = function(child, dict, parent) {
		if (!child || child.getPositionX() == 1 || child.getPositionY() == 1) {
			return;
		}
		
		// standard string
		var scale = _standard_string(dict['Scale']);
		if (!scale || scale.length === 0) {
			return;
		}
		
		var parentSize = parent.getContentSize();
		var childSize = child.getContentSize();
		var scaleX = (parentSize.width * childSize.width == 0) ? 1 : parentSize.width / childSize.width;
		var scaleY = (parentSize.height * childSize.height == 0) ? 1 : parentSize.height / childSize.height;
		
		// replace values
		if (scale.length < 6 || scale.length > 10) {
			// do nothing
		} else if (scale === 'tofill') {
			child.setScale(scaleX, scaleY);
		} else if (scale === 'aspectfill') {
			child.setScale(scaleX > scaleY ? scaleX : scaleY);
		} else if (scale === 'aspectfit') {
			child.setScale(scaleX < scaleY ? scaleX : scaleY);
		} else if (scale === 'aspectx') {
			child.setScale(scaleX);
		} else if (scale === 'aspecty') {
			child.setScale(scaleY);
		}
	};
	
	var addChildren = function(array, spriteOrLayer, node) {
		node = node || this;
		
		var child, zOrder, count = array.length;
		for (var i = 0; i < count; ++i) {
			child = null;
			zOrder = 0;
			
			var data = array[i];
			if (!data || data.length === 0) {
				child = (spriteOrLayer === 0) ? sf.Sprite.create() : sf.Layer.create();
			} else if (typeof data === 'string') {
				cn.log('node from a string: ' + data);
				child = sf.Node.create(data);
			} else if (typeof data === 'object') {
				child = (spriteOrLayer === 0) ? sf.Sprite.create(data) : sf.Layer.create(data);
				// z-order
				zOrder = data['ZOrder'] ? data['ZOrder'] : 0;
				// support scale mode
				processScaleMode(child, data, node);
				// support position adjestment
				processPositionAdjustment(child, data, node);
			} else {
				cn.error(data);
			}
			
			if (child instanceof cc.Node) {
				node.addChild(child, zOrder);
			}
		}
	};
	
	var addSprites = function(array, node) {
		if (array && array.length > 0) {
			addChildren(array, 0, node || this);
		}
	};
	
	var addLayers = function(array, node) {
		if (array && array.length > 0) {
			addChildren(array, 1, node || this);
		}
	};
	
	var getTarget = function(string, node) {
		node = node || this;
		if (!string) {
			return node;
		}
		var target = _standard_string(string);
		
		// target by format: xxx.xxx.xxx
		var pos = target.indexOf('.');
		if (pos > 0) {
			var str1 = target.substring(0, pos);
			var str2 = target.substring(pos + 1);
			return getTarget(str2, getTarget(str1, node));
		}
		
		// target by simple name/tag
		if (target === 'scene') {
			return cn.Node.ancestorScene(node);
		} else if (target === 'self') {
			return node;
		} else if (target === 'parent') {
			return node.getParent();
		} else {
			// children sprite
			var children = node.getChildren();
			for (var i = children.length - 1; i >= 0; --i) {
				if (children[i].getTag() == target) {
					return children[i];
				}
			}
		}
		
		cn.log('no such target: ' + target);
		return null;
	};
	
	//------------------------------------------------------- sf.Node extensions
	
	var instanceExtensions = function(Node) {
		
		Node.prototype.addSprites = addSprites;
		Node.prototype.addLayers = addLayers;
		Node.prototype.getTarget = getTarget;
		
		return Node;
	};
	
	var classExtensions = function(Node) {
		
		Node.addSprites = addSprites;
		Node.addLayers = addLayers;
		Node.getTarget = getTarget;
		
		return Node;
	};
	
	//--------------------------------------------------------------------------
	
	var prop = {
		ctor: Node,
		
		// properties
	};
	
	// extends
	sf.Node = cn.Node.extend(prop);
	
	// static
	sf.Node.create = create;
	
	sf.Node.extensions = instanceExtensions;
	classExtensions(sf.Node);
	
}(SpriteForest);

File: spriteforest-x/base/SFNode.cpp
//
//  SFNode.cpp
//  SpriteForest-X
//
//  Created by Moky on 12-11-24.
//  Copyright 2012 Slanissue Inc. All rights reserved.
//

#include "SFNode.h"

#include "SFSprite.h"
#include "SFLayer.h"
#include "SFScene.h"

NS_SF_BEGIN

USING_NS_CC;
USING_NS_CN;

CCNode * SFNode::nodeWithString(const char * pszString)
{
	if (!pszString || strlen(pszString) == 0)
	{
		CNLog("could not happen!");
		return NULL;
	}
	
	std::string string(pszString);
	string = CNString::trim(string);
	if (CNString::hasPrefix(string, "include class=\""))
	{
		string = string.substr(15);
		std::size_t pos = string.find('"');
		if (pos == std::string::npos)
		{
			CNLog("parse error, string = %s", string.c_str());
			return NULL;
		}
		string = string.substr(0, pos);
	}
	CNLog("creating node from string: %s", string.c_str());
	
	CCObject * pObject = CNRuntimeClass::createObject(string.c_str());
	return dynamic_cast<CCNode *>(pObject);
}

CCNode * SFNode::nodeWithDictionary(CCDictionary * pDict)
{
	if (!pDict)
	{
		CNLog("could not happen!");
		return NULL;
	}
	
	CCString * pClassName = (CCString *)pDict->objectForKey("Class");
	if (!pClassName)
	{
		CNLog("could not happen!");
		return NULL;
	}
	
	CCObject * pObject = CNRuntimeClass::createObject(pClassName->getCString());
	if (SFSprite * pSprite = dynamic_cast<SFSprite *>(pObject))
	{
		pSprite->initWithDictionary(pDict);
		return pSprite;
	}
	else if (SFLayer * pLayer = dynamic_cast<SFLayer *>(pObject))
	{
		pLayer->initWithDictionary(pDict);
		return pLayer;
	}
	else if (SFScene * pScene = dynamic_cast<SFScene *>(pObject))
	{
		pScene->initWithDictionary(pDict);
		return pScene;
	}
	else if (CCNode * pNode = dynamic_cast<CCNode *>(pObject))
	{
		CNLog("user defined node: %s", pClassName->getCString());
		return pNode;
	}
	CNLog("unrecognized node: %s", pClassName->getCString());
	return NULL;
}

std::string & SFNode_stringByRemoveSpaces(std::string & string)
{
//	CNLog("before: [%s]", string.c_str());
	std::size_t p1, p2 = 0;
	while ((p1 = string.find_first_of(" \n\r\t", p2)) != std::string::npos)
	{
		p2 = string.find_first_not_of(" \n\r\t", p1 + 1);
		if (p2 == std::string::npos)
		{
			string = string.substr(0, p1);
			break;
		}
		
		string = string.substr(0, p1) + string.substr(p2);
		p2 = p1 + 1;
	}
//	CNLog(" after: [%s]", string.c_str());
	
	return string;
}

void SFNode_processPositionAdjustment(CCNode * pParent, CCNode * pChild, CCDictionary * pDict)
{
	if (!pChild || abs(pChild->getPositionX()) > 0.01f || abs(pChild->getPositionY()) > 0.01f)
	{
		return;
	}
	CCString * pPosition = (CCString *)pDict->objectForKey("Position");
	std::string position = pPosition ? pPosition->getCString() : "";
	position = SFNode_stringByRemoveSpaces(position);
	if (position.empty())
	{
		return;
	}
	//CNLog("position: %s", position.c_str());
	position = CNString::lowercaseString(position);
	
	// const strings
	if (position.length() < 6 || position.length() > 11)
	{
		// do nothing
	}
	else if (strcmp(position.c_str(), "topleft") == 0)
	{
		position = "{left,top}";
	}
	else if (strcmp(position.c_str(), "topright") == 0)
	{
		position = "{right,top}";
	}
	else if (strcmp(position.c_str(), "center") == 0)
	{
		position = "{center,middle}";
	}
	else if (strcmp(position.c_str(), "bottomleft") == 0)
	{
		position = "{left,bottom}";
	}
	else if (strcmp(position.c_str(), "bottomright") == 0)
	{
		position = "{right,bottom}";
	}
	
	// replacing values
	if (position.length() >= 7) // "{0,top}"
	{
		CCSize parentSize = pParent->getContentSize();
		
		std::string top("0");
		std::string middle("0");
		std::string bottom("0");
		std::string left("0");
		std::string center("0");
		std::string right("0");
		
		top = CNString::stringWithFormat(top, "%.0f", parentSize.height);
		middle = CNString::stringWithFormat(middle, "%.0f", parentSize.height / 2);
		center = CNString::stringWithFormat(center, "%.0f", parentSize.width / 2);
		right = CNString::stringWithFormat(right, "%.0f", parentSize.width);
		
		std::size_t pos;
		
		pos = position.find("top");
		if (pos != std::string::npos)
		{
			position = position.substr(0, pos) + top + position.substr(pos + 3);
		}
		pos = position.find("middle");
		if (pos != std::string::npos)
		{
			position = position.substr(0, pos) + middle + position.substr(pos + 6);
		}
		pos = position.find("bottom");
		if (pos != std::string::npos)
		{
			position = position.substr(0, pos) + bottom + position.substr(pos + 6);
		}
		pos = position.find("left");
		if (pos != std::string::npos)
		{
			position = position.substr(0, pos) + left + position.substr(pos + 4);
		}
		pos = position.find("center");
		if (pos != std::string::npos)
		{
			position = position.substr(0, pos) + center + position.substr(pos + 6);
		}
		pos = position.find("right");
		if (pos != std::string::npos)
		{
			position = position.substr(0, pos) + right + position.substr(pos + 5);
		}
	}
	//CNLog("position: %s", position.c_str());
	
	// set it
	if (position.length() >= 5) // "{0,0}"
	{
		pChild->setPosition(CCPointFromString(position.c_str()));
	}
}

void SFNode_processScaleMode(CCNode * pParent, CCNode * pChild, CCDictionary * pDict)
{
	if (!pChild || abs(pChild->getScaleX() - 1) > 0.01f || abs(pChild->getScaleY() - 1) > 0.01f)
	{
		return;
	}
	CCString * pScale = (CCString *)pDict->objectForKey("Scale");
	std::string scale = pScale ? pScale->getCString() : "";
	scale = SFNode_stringByRemoveSpaces(scale);
	if (scale.empty())
	{
		return;
	}
	//CNLog("scale: %s", scale.c_str());
	scale = CNString::lowercaseString(scale);
	
	CCSize parentSize = pParent->getContentSize();
	CCSize childSize = pChild->getContentSize();
	float fScaleX = (parentSize.width * childSize.width == 0) ? 1 : parentSize.width / childSize.width;
	float fScaleY = (parentSize.height * childSize.height == 0) ? 1 : parentSize.height / childSize.height;
	
	if (scale.length() < 6 || scale.length() > 10)
	{
		// do nothing
	}
	else if (strcmp(scale.c_str(), "tofill") == 0)
	{
		pChild->setScaleX(fScaleX);
		pChild->setScaleY(fScaleY);
	}
	else if (strcmp(scale.c_str(), "aspectfill") == 0)
	{
		pChild->setScale(MAX(fScaleX, fScaleY));
	}
	else if (strcmp(scale.c_str(), "aspectfit") == 0)
	{
		pChild->setScale(MIN(fScaleX, fScaleY));
	}
	else if (strcmp(scale.c_str(), "aspectx") == 0)
	{
		pChild->setScale(fScaleX);
	}
	else if (strcmp(scale.c_str(), "aspecty") == 0)
	{
		pChild->setScale(fScaleY);
	}
}

void SFNode_addChildren(CCNode * pNode, CCArray * pArray, const int spriteOrLayer)
{
	CCNode * pChild;
	int iZOrder;
	
	int count = pArray->count();
	for (int i = 0; i < count; i++)
	{
		CCPoolManager * pm = CCPoolManager::sharedPoolManager();
		pm->push();
		
		pChild = NULL;
		iZOrder = 0;
		
		CCObject * pData = pArray->objectAtIndex(i);
		if (CCString * pString = dynamic_cast<CCString *>(pData))
		{
			CNLog("node from a string: %s", pString->getCString());
			pChild = (spriteOrLayer == 0) ? SFSprite::nodeWithString(pString->getCString()) : SFLayer::nodeWithString(pString->getCString());
		}
		else if (CCDictionary * pDict = dynamic_cast<CCDictionary *>(pData))
		{
			pChild = (spriteOrLayer == 0) ? SFSprite::nodeWithDictionary(pDict) : SFLayer::nodeWithDictionary(pDict);
			CCString * pZOrder = (CCString *)pDict->objectForKey("ZOrder");
			iZOrder = (pZOrder) ? pZOrder->intValue() : 0;
			
			// support Scale Mode
			SFNode_processScaleMode(pNode, pChild, pDict);
			
			// support Position Adjustment
			SFNode_processPositionAdjustment(pNode, pChild, pDict);
		}
		else
		{
			CNLog("unrecongnized node");
		}
		
		if (pChild)
		{
			pNode->addChild(pChild, iZOrder);
		}
		else
		{
			CNLog("error");
		}
		
		pm->pop();
	}
}

void SFNode::addSprites(CCNode * pNode, CCArray * pArray)
{
	if (pNode && pArray)
	{
		SFNode_addChildren(pNode, pArray, 0);
	}
}

void SFNode::addLayers(CCNode * pNode, CCArray * pArray)
{
	if (pNode && pArray)
	{
		SFNode_addChildren(pNode, pArray, 1);
	}
}

CCNode * SFNode::getTarget(CCNode * pNode, const char * pszTarget)
{
	if (!pNode || !pszTarget || strlen(pszTarget) == 0)
	{
		return pNode;
	}
	
	// target by format: xxx.xxx.xxx
	std::string str(pszTarget);
	int pos = str.find('.');
	if (pos != std::string::npos)
	{
		std::string str1(str.substr(0, pos));
		std::string str2(str.substr(pos + 1));
		return getTarget(getTarget(pNode, str1.c_str()), str2.c_str());
	}
	
	// target by simple name/tag
	if (strcasecmp(pszTarget, "scene") == 0)
	{
		return CNNode::ancestorScene(pNode);
	}
	else if (strcasecmp(pszTarget, "self") == 0)
	{
		return pNode;
	}
	else if (strcasecmp(pszTarget, "parent") == 0)
	{
		return pNode->getParent();
	}
	else
	{
		// children sprite
		int iTag = atoi(pszTarget);
		CCObject * pChild = NULL;
		CCARRAY_FOREACH_REVERSE(pNode->getChildren(), pChild)
		{
			if (((CCNode *)pChild)->getTag() == iTag)
			{
				return (CCNode *)pChild;
			}
		}
	}
	
	CNLog("no such target: %s", pszTarget);
	return NULL;
}

NS_SF_END
